File: app/src/main/java/me/zhanghai/android/files/provider/remote/RemotePathObservable.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.remote;

import android.os.Parcel;
import android.os.Parcelable;
import android.os.RemoteException;

import java.io.IOException;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import me.zhanghai.android.files.provider.common.PathObservable;
import me.zhanghai.android.files.util.RemoteCallback;

public class RemotePathObservable implements PathObservable, Parcelable {

    @Nullable
    private final PathObservable mLocalPathObservable;

    @Nullable
    private final IRemotePathObservable mRemotePathObservable;

    private final Set<Runnable> mRemoteObservers;

    private boolean mRemoteInitialized;

    private final Object mRemoteLock;

    public RemotePathObservable(@NonNull PathObservable pathObservable) {
        mLocalPathObservable = pathObservable;
        mRemotePathObservable = null;
        mRemoteObservers = null;
        mRemoteLock = null;
    }

    public void initForRemote() throws IOException {
        synchronized (mRemoteLock) {
            if (mRemoteInitialized) {
                throw new IllegalStateException();
            }
            try {
                mRemotePathObservable.addObserver(new RemoteCallback(result -> {
                    synchronized (mRemoteLock) {
                        for (Runnable observer : mRemoteObservers) {
                            observer.run();
                        }
                    }
                }));
            } catch (RemoteException e) {
                close();
                throw new RemoteFileSystemException(e);
            }
            mRemoteInitialized = true;
        }
    }

    @Override
    public void addObserver(@NonNull Runnable observer) {
        Objects.requireNonNull(observer);
        synchronized (mRemoteLock) {
            if (!mRemoteInitialized) {
                throw new IllegalStateException();
            }
            mRemoteObservers.add(observer);
        }
    }

    @Override
    public void removeObserver(@NonNull Runnable observer) {
        Objects.requireNonNull(observer);
        synchronized (mRemoteLock) {
            if (!mRemoteInitialized) {
                throw new IllegalStateException();
            }
            mRemoteObservers.remove(observer);
        }
    }

    @Override
    public void close() throws IOException {
        if (mRemotePathObservable != null) {
            synchronized (mRemoteLock) {
                ParcelableException exception = new ParcelableException();
                try {
                    mRemotePathObservable.close(exception);
                } catch (RemoteException e) {
                    throw new RemoteFileSystemException(e);
                }
                exception.throwIfNotNull();
                mRemoteObservers.clear();
            }
        } else {
            mLocalPathObservable.close();
        }
    }

    private static class Stub extends IRemotePathObservable.Stub {

        @NonNull
        private final PathObservable mPathObservable;

        public Stub(@NonNull PathObservable pathObservable) {
            mPathObservable = pathObservable;
        }

        @Override
        public void addObserver(@NonNull RemoteCallback observer) {
            Objects.requireNonNull(observer);
            mPathObservable.addObserver(() -> observer.sendResult(null));
        }

        @Override
        public void close(@NonNull ParcelableException exception) {
            try {
                mPathObservable.close();
            } catch (IOException | RuntimeException e) {
                exception.set(e);
            }
        }
    }


    public static final Creator<RemotePathObservable> CREATOR =
            new Creator<RemotePathObservable>() {
                @Override
                public RemotePathObservable createFromParcel(Parcel source) {
                    return new RemotePathObservable(source);
                }
                @Override
                public RemotePathObservable[] newArray(int size) {
                    return new RemotePathObservable[size];
                }
            };

    protected RemotePathObservable(Parcel in) {
        mLocalPathObservable = null;
        mRemotePathObservable = IRemotePathObservable.Stub.asInterface(
                in.readStrongBinder());
        mRemoteObservers = new HashSet<>();
        mRemoteLock = new Object();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        if (mRemotePathObservable != null) {
            throw new IllegalStateException("Already at the remote side");
        }
        dest.writeStrongBinder(new Stub(mLocalPathObservable).asBinder());
    }
}

File: app/src/main/java/me/zhanghai/android/files/provider/remote/RemotePathObservable.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.remote

import android.os.Bundle
import android.os.Parcel
import android.os.Parcelable
import android.os.RemoteException
import me.zhanghai.android.files.provider.common.PathObservable
import me.zhanghai.android.files.util.RemoteCallback
import java.io.IOException

class RemotePathObservable : PathObservable, Parcelable {
    private val localPathObservable: PathObservable?
    private val remotePathObservable: IRemotePathObservable?
    private val remoteObservers: MutableSet<() -> Unit>?
    private var isRemoteInitialized = false
    private val remoteLock: Any?

    constructor(pathObservable: PathObservable) {
        localPathObservable = pathObservable
        remotePathObservable = null
        remoteObservers = null
        remoteLock = null
    }

    @Throws(IOException::class)
    fun initializeForRemote() {
        synchronized(remoteLock!!) {
            check(!isRemoteInitialized)
            try {
                remotePathObservable!!.addObserver(RemoteCallback {
                    synchronized(remoteLock) { remoteObservers!!.forEach { it() } }
                })
            } catch (e: RemoteException) {
                close()
                throw RemoteFileSystemException(e)
            }
            isRemoteInitialized = true
        }
    }

    override fun addObserver(observer: () -> Unit) {
        synchronized(remoteLock!!) {
            check(isRemoteInitialized)
            remoteObservers!!.add(observer)
        }
    }

    override fun removeObserver(observer: () -> Unit) {
        synchronized(remoteLock!!) {
            check(isRemoteInitialized)
            remoteObservers!!.remove(observer)
        }
    }

    @Throws(IOException::class)
    override fun close() {
        if (remotePathObservable != null) {
            synchronized(remoteLock!!) {
                remotePathObservable.call { exception -> close(exception) }
                remoteObservers!!.clear()
            }
        } else {
            localPathObservable!!.close()
        }
    }

    private class Stub(private val pathObservable: PathObservable) : IRemotePathObservable.Stub() {
        override fun addObserver(observer: RemoteCallback) {
            pathObservable.addObserver { observer.sendResult(Bundle()) }
        }

        override fun close(exception: ParcelableException) {
            tryRun(exception) { pathObservable.close() }
        }
    }

    private constructor(source: Parcel) {
        localPathObservable = null
        remotePathObservable = IRemotePathObservable.Stub.asInterface(source.readStrongBinder())
        remoteObservers = mutableSetOf()
        remoteLock = Any()
    }

    override fun describeContents(): Int = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        check(remotePathObservable == null) { "Already at the remote side" }
        dest.writeStrongBinder(Stub(localPathObservable!!).asBinder())
    }

    companion object {
        @JvmField
        val CREATOR = object : Parcelable.Creator<RemotePathObservable> {
            override fun createFromParcel(source: Parcel): RemotePathObservable =
                RemotePathObservable(source)

            override fun newArray(size: Int): Array<RemotePathObservable?> = arrayOfNulls(size)
        }
    }
}

fun PathObservable.toRemote(): RemotePathObservable = RemotePathObservable(this)
